lib/src/main/java/io/ably/lib/realtime/Channel.java

```diff
@@ -94,12 +94,51 @@ private void setState(ChannelState newState, ErrorInfo reason) {
 	 * @throws AblyException
 	 */
 	public void attach() throws AblyException {
+		attach(null);
+	}
+
+	/**
+	 * Attach to this channel.
+	 * This call initiates the attach request, and the response
+	 * is indicated asynchronously in the resulting state change.
+	 * attach() is called implicitly when publishing or subscribing
+	 * on this channel, so it is not usually necessary for a client
+	 * to call attach() explicitly.
+	 *
+	 * @param listener When the channel is attached successfully or the attach fails and
+	 * the ErrorInfo error is passed as an argument to the callback
+	 * @throws AblyException
+	 */
+	public void attach(final CompletionListener listener) throws AblyException {
 		Log.v(TAG, "attach(); channel = " + name);
 		/* check preconditions */
 		switch(state) {
 			case attaching:
+				if (listener != null) {
+					on(new ChannelStateListener() {
```

Comment:
Can't we remove this duplication by having a `private static class AttachListener implements ChannelStateListener { .... }` ?